[Clang] Improve error for `-fsanitize=function/kcfi -mexecute-only` incompatibility

Il-Capitano opened the pull request:
The current error message when using the `-fsanitize=function -mexecute-only` flags together points to the target triple as the reason that `-fsanitize=function` is not allowed to be used, even when the function sanitizer is otherwise supported on the target when not using `-mexecute-only`.

The error message is improved to give `-mexecute-only` as the reason for disallowing `-fsanitize=function` if it was passed to the driver.

Fixes https://github.com/llvm/llvm-project/issues/117974

Il-Capitano committed:
[Clang] Improve error for `-fsanitize=function/kcfi -mexecute-only` incompatibility

The current error message when using the `-fsanitize=function -mexecute-only`
flags together points to the target triple as the reason that
`-fsanitize=function` is not allowed to be used, even when the function
sanitizer is otherwise supported on the target when not using
`-mexecute-only`.

The error message is improved to give `-mexecute-only` as the reason for
disallowing `-fsanitize=function` if it was passed to the driver.

Il-Capitano commented on clang/lib/Driver/SanitizerArgs.cpp:
```diff
@@ -418,8 +418,14 @@ SanitizerArgs::SanitizerArgs(const ToolChain &TC,
+                  << Desc << A->getAsString(Args);
+            else
+              D.Diag(diag::err_drv_argument_not_allowed_with)
+                  << Desc << Triple.str();
```
Maybe `diag::err_drv_unsupported_opt_for_target` would be a better diagnostic in this case? That would produce
```
error: unsupported option '-fsanitize=function' for target '<triple>'
```

Il-Capitano commented:
@MaskRay Can I ask you to review/add appropriate reviewers for this patch?

MaskRay reviewed (approved):
Yes, err_drv_unsupported_opt_for_target will be better for the incompatible triple error

Il-Capitano committed:
Use `err_drv_unsupported_opt_for_target` and add test for `-mpure-code`

Il-Capitano commented:
Thanks for reviewing!

I updated the code to use `err_drv_unsupported_opt_for_target`, and added test cases for the `-mpure-code` alias of `-mexecute-only` also.
